src/sensai/clustering/base/clustering.py

```diff
@@ -0,0 +1,196 @@
+            percentageGenerated = int(100 * i / self.numClusters)
+            if percentageGenerated == percentageToLog:
+                log.info(f"Processed {percentageToLog}% of clusters")
+                percentageToLog += 5
+
+            cluster = self.getCluster(clusterId)
+            if condition is None or condition(cluster):
+                yield cluster
+
+    def noiseCluster(self):
+        if self.noiseLabel is None:
+            raise NotImplementedError(f"The algorithm {self} does not provide a noise cluster")
+        return self.getCluster(self.noiseLabel)
+
+    def summaryDF(self, condition: Callable[[Cluster], bool] = None):
+        """
+        Data frame containing coarse information about the clusters
+
+        :param condition: if provided, only clusters fulfilling the condition will be included
+        :return: pandas DataFrame
+        """
+        summary_dicts = [cluster.summaryDict() for cluster in self.clusters(condition=condition)]
+        return pd.DataFrame(summary_dicts).set_index("identifier", drop=True)
+
+    def fit(self, data: np.ndarray) -> None:
+        log.info(f"Fitting {self} to {len(data)} coordinate datapoints.")
+        self._fit(data)
+        self._datapoints = data
+        self._labels = self.getLabels()
+        if len(self._labels) != len(data):
+            raise Exception(f"Number of labels does not match number of datapoints")
+        self._clusterIdentifiers = set(self._labels)
+        log.info(f"{self} found {self.numClusters} clusters")
+
+    @property
+    def isFitted(self):
+        return self._datapoints is not None
+
+    @property
+    def datapoints(self) -> np.ndarray:
+        assert self.isFitted
+        return self._datapoints
+
+    def getLabels(self) -> np.ndarray:
+        assert self.isFitted
+        labels = self._getLabels()
+        # Relabel clusters that do not fulfill size bounds as noise
+        if self.minClusterSize != -np.inf or self.maxClusterSize != np.inf:
+            for clusterId, clusterSize in zip(*np.unique(labels, return_counts=True)):
+                if not self.minClusterSize <= clusterSize <= self.maxClusterSize:
+                    labels[labels == clusterId] = self.noiseLabel
+        return labels
+
+    @property
+    def clusterIdentifiers(self) -> Set[int]:
+        assert self.isFitted
+        return self._clusterIdentifiers
+
+    # unfortunately, there seems to be no way to annotate the return type correctly
+    # https://github.com/python/mypy/issues/3993
+    def getCluster(self, clusterId: int) -> Cluster:
+        if clusterId not in self.getLabels():
+            raise KeyError(f"no cluster for id {clusterId}")
+        return self.Cluster(self.datapoints[self.getLabels() == clusterId], identifier=clusterId)
+
+    @property
+    def numClusters(self) -> int:
+        return len(self.clusterIdentifiers.difference({self.noiseLabel}))
+
+    @abstractmethod
+    def _fit(self, x: np.ndarray) -> None:
```

Comment:
This is quite a strange interface. The purpose of this function seems to be to produce a side-effect that achieves that labels can subsequently be retrieved via _getLabels. Wouldn't a more fitting interface (pun intended) make the function _fit return whatever _getLabels returns?
We could give the function a clearer name, too (_computeLabels).

Reply:
Prefer functionally defined interfaces whenever possible.

Reply:
I wrote the class around the scikit-learn clustering algorithms which works exactly in the way you described, this made the SKLearnClusteringModel implementation very natural. But I agree, it will makes things easier to let _fit return the labels. I'll adjust it

Reply:
Done, thanks for the comment! It's much cleaner now